### PR TITLE
SOAR-13594,13667: Update Zoom Auth and Add JWT

### PR DESCRIPTION
**Pull request by mrinehart-r7**
### Description

Describe the proposed changes:

  - Improves OAuth authentication logic to help mitigate infinite looping
  - Adds back JWT authentication to possible connection configurations until OAuth token invalidation is fully solved. Please note: JWT auth will become unavailable in June 2023.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Testing

Task with JWT Authentication
Screen Shot 2023-04-05 at 4 50 35 PM

Task with OAuth Authentication
Screen Shot 2023-04-05 at 4 50 52 PM

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [x] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [x] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [x] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [x] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [x] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [x] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [x] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [x] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/zoom/icon_zoom/util/api.py [hunk 1]
```diff
@@ -90,19 +67,67 @@ def get_user_activity_events(
             else:
                 return events
 
+    def _refresh_oauth_token(self):
```
Comment (mrinehart-r7) [inline plugins/zoom/icon_zoom/util/api.py:70]: This function will either fail entirely without any retries or succeed and does not re-use the _call_api function, which should eliminate any chance of getting into a retry loop.

File: plugins/zoom/icon_zoom/util/api.py [hunk 2]
```diff
@@ -167,22 +225,29 @@ def _handle_response(self, response: Response, allow_404: bool, original_call_ar
...
+    def _handle_rate_limit_error(response: Response) -> PluginException:
         rate_limit_type = response.headers.get("X-RateLimit-Type", "")
         rate_limit_limit = response.headers.get("X-RateLimit-Limit")
         rate_limit_remaining = response.headers.get("X-RateLimit-Remaining")
         rate_limit_retry_after = response.headers.get("Retry-After")
 
         if rate_limit_type == "Light":
-            raise PluginException(
+            return PluginException(
                 cause="Account is rate-limited by the maximum per-second limit for this API.",
```
Comment (dlaverty-r7) [inline plugins/zoom/icon_zoom/util/api.py:236]: Just wondering if we hit the max per-second limit for the API should we wait and retry after the specified interval?

Comment (mrinehart-r7) [reply plugins/zoom/icon_zoom/util/api.py:236]: I talked with @dgreene-r7 about this and he wants us to just raise an exception.

File: plugins/zoom/icon_zoom/util/api.py [hunk 3]
```diff
@@ -167,22 +225,29 @@ def _handle_response(self, response: Response, allow_404: bool, original_call_ar
...
             )
+
+        return PluginException(cause="Account is rate-limited by an unknown quota.", assistance="Try again later.")
+
+    def _is_using_oauth(self) -> bool:
+        if (self.account_id and self.client_id and self.client_secret) or self.oauth_token:
+            return True
         else:
-            raise PluginException(cause="Account is rate-limited by an unknown quota.", assistance="Try again later.")
+            return False
```
[the remark this answers is not in the document]
Comment (mrinehart-r7) [reply plugins/zoom/icon_zoom/util/api.py:253]: I thought about that initially but I think this version is more readable.

**Review by tcheng-r7 — APPROVED** [review]

File: plugins/zoom/icon_zoom/util/api.py [hunk 4]
```diff
@@ -142,23 +192,31 @@ def _handle_response(self, response: Response, allow_404: bool, original_call_ar
             raise PluginException(preset=PluginException.Preset.BAD_REQUEST, data=response.json())
 
         if response.status_code == 401:
-            self.logger.info(f"Received HTTP {response.status_code}, re-authenticating to Zoom...")
-            self.refresh_oauth_token()
-            return self._call_api(**original_call_args)
+            if self.oauth_token or self._is_using_oauth():
+                self.logger.info(f"Received HTTP {response.status_code}, re-authenticating to Zoom...")
+                self._refresh_oauth_token()
+                return self._call_api(**original_call_args)
```
Comment (dgreene-r7) [inline plugins/zoom/icon_zoom/util/api.py:198]: I feel like we might need a call counter here 'cause we could end up in an infinite loop of trying to request a new token, getting a 401 back, and then trying again forever.

**Review by gusingh-r7 — APPROVED** [review]

File: plugins/zoom/icon_zoom/util/api.py [hunk 5]
```diff
@@ -24,39 +24,27 @@ def __call__(self, request: requests.Request):
...
 class ZoomAPI:
-    def __init__(self, account_id: str, client_id: str, client_secret: str, logger: Logger):
+    def __init__(
+        self,
+        logger: Logger,
+        account_id: Optional[str] = None,  # For OAuth only
+        client_id: Optional[str] = None,  # For OAuth only
+        client_secret: Optional[str] = None,  # For OAuth only
+        jwt_token: Optional[str] = None,
+    ):  # For JWT auth only
```
[the remark this answers is not in the document]
Comment (mrinehart-r7) [reply plugins/zoom/icon_zoom/util/api.py:34]: Moved it back - hopefully Black doesn't fight me to move it back down to the newline again 😝

**Review by dlaverty-r7 — APPROVED** [review]

**Review by dgreene-r7 — APPROVED** [review]

**Pull request merged**